src/main/java/org/jboss/threads/JBossExecutors.java

```diff
@@ -336,10 +336,15 @@ static ClassLoader getContextClassLoader(final Thread thread) {
      * @return the old context class loader
      */
     static ClassLoader getAndSetContextClassLoader(final Thread thread, final ClassLoader newClassLoader) {
+        boolean same = false;
         try {
-            return getContextClassLoader(thread);
+            final ClassLoader cl = getContextClassLoader(thread);
+            same = (cl == newClassLoader);
+            return cl;
         } finally {
-            setContextClassLoader(thread, newClassLoader);
+            if (!same) {
+                setContextClassLoader(thread, newClassLoader);
+            }
         }
```

Comment:
I wonder if we'd be better off removing the try/finally here?

```java
final ClassLoader cl = getContextClassLoader(thread);
if (newClassLoader != newClassLoader) setContextClassLoader(thread, newClassLoader);
return cl;
```

Reply:
yeah I believe is better but maybe there' was a reason for it? I have really no idea!

Reply:
I'm still in thinking mode here!

Reply:
I'd drop the `try/finally` - it almost certainly adds nothing, but it does add complexity.